src/inspector_js_api.cc

```diff
@@ -32,6 +32,171 @@ std::unique_ptr<StringBuffer> ToProtocolString(Isolate* isolate,
+    return result.ToLocal(&r) && r->IsTrue();
+  }
+
+  void SetCallback(Local<Function> callback) {
+    callback_.Reset(env_->isolate(), callback);
+  }
+
+  void SetDispatcher(JSDispatcher* dispatcher) {
+    if (dispatcher == nullptr && dispatcher_ != nullptr) {
+      HandleMessage(v8_inspector::StringView());
+    }
+    dispatcher_ = dispatcher;
+  }
+
+  size_t self_size() const override {
+    return sizeof(*this);
+  }
+
+  static void SendMessageToFrontend(const FunctionCallbackInfo<Value>& info);
+
+ private:
+  Environment* env_;
+  JSDispatcher* dispatcher_;
+  Persistent<Function> callback_;
+};
+
+class JSDispatcher : public MessageDispatcher {
+ public:
+  JSDispatcher(std::unique_ptr<DispatcherSession>&& session,
+               JSDispatcherInterface* interface)
+               : js_interface_(interface), session_(std::move(session)) {
+    interface->SetDispatcher(this);
+  }
+
+  ~JSDispatcher() override {
+    if (js_interface_ != nullptr) {
+      js_interface_->SetDispatcher(nullptr);
+    }
+  }
+
+  bool HandleMessage(const v8_inspector::StringView& message) override {
+    return js_interface_ != nullptr && js_interface_->HandleMessage(message);
+  }
+
+  void ResetInterface() {
+    js_interface_ = nullptr;
+  }
+
+  void SendMessageToFrontend(const StringView& message) {
+    session_->SendMessageToFrontend(message);
+  }
+
+ private:
+  JSDispatcherInterface* js_interface_;
+  std::unique_ptr<DispatcherSession> session_;
+};
+
+JSDispatcherInterface::~JSDispatcherInterface() {
+  callback_.Reset();
+  if (dispatcher_ != nullptr) {
+    dispatcher_->ResetInterface();
+  }
+}
+
+void JSDispatcherInterface::SendMessageToFrontend(
+    const FunctionCallbackInfo<Value>& info) {
+  JSDispatcherInterface* interface;
+  ASSIGN_OR_RETURN_UNWRAP(&interface, info.Holder());
+  Environment* env = interface->env_;
+  JSDispatcher* dispatcher = interface->dispatcher_;
+  if (info.Length() != 1 || !info[0]->IsString()) {
+    env->ThrowTypeError("Protocol message should be a string");
+    return;
+  }
+  if (dispatcher == nullptr) {
+    return;
+  }
+  dispatcher->SendMessageToFrontend(
+      ToProtocolString(env->isolate(), info[0])->string());
+}
+
+class JSDispatcherFactory : public MessageDispatcherFactory {
+ public:
+  JSDispatcherFactory(Environment* env, Local<Function> js_factory)
+                      : env_(env), js_factory_(env->isolate(), js_factory) {}
+
+  std::unique_ptr<MessageDispatcher> Create(
+      std::unique_ptr<DispatcherSession> session)
+      override {
+    Isolate* isolate = env_->isolate();
+    Local<Context> context = env_->context();
+    HandleScope handle_scope(isolate);
+    Context::Scope context_scope(context);
+
+    Local<FunctionTemplate> tmpl = env_->NewFunctionTemplate(nullptr);
+    tmpl->InstanceTemplate()->SetInternalFieldCount(1);
+    tmpl->SetClassName(
+        FIXED_ONE_BYTE_STRING(env_->isolate(),
+                              "InspectorDispatcherConnection"));
```

Comment:
Instead of creating such a template every single time, why not cache it in the Environment?